### PR TITLE
observables: make observables immutable.

### LINKED ISSUE #4208: make observables immutable

### PR DESCRIPTION
**Pull request by KaiSzuttor**
Fixes #4208 

Description of changes:
- made observables immutable

ToDo:
- fix issue in python tests of cylindrical observables

### CODE REVIEW

File: src/core/observables/PidObservable.hpp [hunk 1]
```diff
@@ -51,7 +51,7 @@ using ParticleReferenceRange =
  */
 class PidObservable : virtual public Observable {
   /** Identifiers of particles measured by this observable */
-  std::vector<int> m_ids;
+  const std::vector<int> m_ids;
```
Comment (fweik) [inline src/core/observables/PidObservable.hpp]: I think, as a general rule, const data members should be avoided.

Comment (KaiSzuttor) [reply src/core/observables/PidObservable.hpp]: ok, changed that

File: src/core/observables/ProfileObservable.hpp [hunk 2]
```diff
@@ -34,25 +34,37 @@ namespace Observables {
 
 /** Cartesian profile observable */
 class ProfileObservable : virtual public Observable {
+protected:
+  /** Range of the profile edges. */
+  const std::array<std::pair<double, double>, 3> limits;
```
Comment (fweik) [inline src/core/observables/ProfileObservable.hpp]: Here ...

File: src/core/observables/ProfileObservable.hpp [hunk 3]
```diff
@@ -34,25 +34,37 @@ namespace Observables {
...
+  /** Number of bins for each coordinate. */
+  const std::array<size_t, 3> n_bins;
```
Comment (fweik) [inline src/core/observables/ProfileObservable.hpp]: ... too.

There reason is that this limits how the type can be used, which breaks the separation, but has no actual benefit. Also `protected` is usually no a great sign, design-wise. This is even a core guideline.

**Review by jngrad — DISMISSED** [review]
LGTM

**Review by jngrad — APPROVED** [review]

**Pull request merged**